### DISCUSSION TITLE
Generic/templated types and/or record-oriented functions that don’t need precise type-matches

### DISCUSSION THREAD

**Discussion by MP-Corelight** [zeek/zeek#3841] [Ideas]
As a writer of Zeek Scripts, I'd like a little more flexibility in code re-use — e.g. functions that vary slightly in terms of data types shouldn't need to be copy/pasted in their entirety with minor changes. There are some parallels with templates and macros, from C++. The biggest frustration is that this leads to a lot of duplicate code.

**Comment by awelzel** [comment]
> or record-oriented functions that don’t need precise type-matches

I was wondering about @JustinAzoff 's thoughts over at https://github.com/zeek/zeek/issues/3805#issuecomment-2221257889

This is not templating, nor OOO and might look somewhat duck-typish-backwards, but hey, would be curious what people think.

What if `a as T` was extended to `a as record T` that only works for record values and returns a value that has record type `T`, but internally keeps a reference to the original record value. Field lookups are translated from this value of type `T` to the original value. This translation/wrapping/proxying would only work if `T` has a matching subset of fields, otherwise can fail statically.

For values of type `any` - like with the `log_stream_policy` example from in the comment - dynamically checking should still be efficient if we cache involved type identifiers.

```
type Common: record {
  c: count;
  s: string &optional;
};

type R: record {
  extra: string;
  s: string &optional;
  c: count;
  t: time;
};

function work_with_common(c: Common) {
  ...
}


local r = R(...);
local cr = r as record Common;
work_with_common(cr);

local ra: any = ...; # from somewhere
local cra = ra as record Common;  # ra has matching fields, succeeds
print cra$c;
```

Would that be overly strange and quirky?

**Reply by vpax** [reply]
I view that as a bit overly strange and quirky 😏. The way I'm thinking about this, it should Just Work to call `work_with_common(r)`. Under the hood, this would (1) type-check that `r` contains all of the fields present in `Common`, with the same names and types, (2) generate a new internal version of `work_with_common()` that takes an `R` record as its argument rather than a `Common` record, (3) expand the original call to instead be to that internal version. The plus of doing it this way, rather than what you sketch above, is both less clunky expression, and there's no need to translate record offsets at run-time.

For the `any` example, I'd be fine with requiring a type-switch (or an `?as` per the discussion that you linked to above) to turn it into either an `R` or a `Common` if it's indeed exactly one or the other; and with not matching either of those if it's not.

**Comment by bbannier** [comment]
@MP-Corelight I think it would be helpful to collect a list of compact, concrete use cases. This would better allow to properly discuss and weigh different design approaches.

There is nothing keeping users from using any [existing](https://jinja.palletsprojects.com) [templating](https://docs.ruby-lang.org/en/master/ERB.html) [engine](https://en.wikipedia.org/wiki/C_preprocessor) to generate Zeek script code, so I suspect the push to get support for this in Zeek script itself comes from some need for deeper integration which is not yet spelled out.

I am unsure whether above discussion about `Common` is all we need to keep in mind though it is definitely an interesting discussion whether we'd want to have type checking at load time or runtime time ("duck typing"). I think having concrete use cases would help understand better what our options are and where trade-offs lie.

**Reply by MP-Corelight** [reply]
We had a real-time discussion on this issue today. It sounds like we can pull some good examples of the need from Corelight's work, so that's covered. The next step would be to construct the requirement in such a way that we can solve it simply and without getting into wildly complex things (a la the direction C++ has taken in some ways) — which is roughly what @vpax has described above, I believe. @ckreibich mentioned he'd like to dig in a little further.

**Comment by MP-Corelight** [comment]
Progress: @vpax and @ckreibich will sit down and sketch out a likely solution before implementation begins, just to avoid late objections.
